src/shared/messaging/test/port-rpc-test.js

```diff
@@ -175,23 +175,158 @@ describe('PortRPC', () => {
     assert.calledWith(testMethod, 'second', 'call');
   });
 
-  describe('#disconnect', () => {
-    it('disconnects and returns port', async () => {
-      const rpc = new PortRPC();
-      const testMethod = sinon.stub();
-      rpc.on('test', testMethod);
+  it('should send "connect" event after connection', async () => {
+    const { port1, port2 } = new MessageChannel();
+    const sender = new PortRPC();
+    const receiver = new PortRPC();
+    const connectHandler = sinon.stub();
+
+    receiver.on('connect', connectHandler);
+    receiver.connect(port2);
+    sender.connect(port1);
+
+    await waitForMessageDelivery();
+
+    assert.calledWith(connectHandler);
+  });
+
+  it('should send "close" event after port is destroyed', async () => {
+    const { port1, port2 } = new MessageChannel();
+    const sender = new PortRPC();
+    const receiver = new PortRPC();
+    const closeHandler = sinon.stub();
+
+    receiver.on('close', closeHandler);
+    receiver.connect(port2);
+    sender.connect(port1);
+    await waitForMessageDelivery();
+
+    closeHandler.resetHistory();
+    sender.destroy();
+    await waitForMessageDelivery();
+
+    assert.calledWith(closeHandler);
+  });
+
+  it('should send "close" event when window is unloaded', async () => {
+    const { port1, port2 } = new MessageChannel();
+    const sender = new PortRPC();
+    const receiver = new PortRPC();
+    const closeHandler = sinon.stub();
 
-      assert.isNull(rpc.disconnect());
+    receiver.on('close', closeHandler);
+    receiver.connect(port2);
+    sender.connect(port1);
+    await waitForMessageDelivery();
 
-      const { port1 } = new MessageChannel();
-      rpc.connect(port1);
+    closeHandler.resetHistory();
+    window.dispatchEvent(new Event('unload'));
+    await waitForMessageDelivery();
 
-      assert.equal(rpc.disconnect(), port1);
-      assert.isNull(rpc.disconnect());
+    assert.calledWith(closeHandler);
+  });
 
-      rpc.call('test');
+  it('should only invoke "close" handler once', async () => {
+    const { port1, port2 } = new MessageChannel();
+    const sender = new PortRPC();
+    const receiver = new PortRPC();
+    const closeHandler = sinon.stub();
+
+    receiver.on('close', closeHandler);
+    receiver.connect(port2);
+    sender.connect(port1);
+
+    // Invoke "close" manually. In a real app it will be invoked when the
+    // window is unloaded and/or the PortRPC is destroyed.
+    sender.call('close');
+    sender.call('close');
+
+    await waitForMessageDelivery();
+
+    assert.calledOnce(closeHandler);
+  });
+
+  /** Transfer a MessagePort to another frame and return the transferred port. */
+  async function transferPort(port, targetWindow) {
+    const transferredPort = new Promise(resolve => {
+      targetWindow.addEventListener('message', e => {
```

Comment:
That's interesting... setting an event listener in another window from the current window. And then drive the port from the current window. 🤯

Reply:
I tried a few different approaches before I landed on the one currently used in this PR. Hopefully this will be a useful reference for some other test in future.